hipercow-api/Tools/ClusterHandleCache.cs

```diff
@@ -1,34 +1,95 @@
     /// it every time (which is slow).
     /// </summary>
-    public class ClusterHandleCache : Dictionary<string, HipercowScheduler>
+    public class ClusterHandleCache : Dictionary<string, IHipercowScheduler>
     {
+        private static ClusterHandleCache? singletonClusterHandleCache;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="ClusterHandleCache"/> class.
         /// </summary>
         public ClusterHandleCache()
         {
-            new Dictionary<string, HipercowScheduler>();
+            new Dictionary<string, IHipercowScheduler>();
+        }
+
+        /// <summary>
+        /// Return the singleton cluster handle cache.
+        /// </summary>
+        /// <returns>The static cluster handle cache for the API.</returns>
+        public static ClusterHandleCache GetSingletonClusterHandleCache()
+        {
+            if (singletonClusterHandleCache == null)
+            {
+                singletonClusterHandleCache = new ClusterHandleCache();
+            }
+
+            return singletonClusterHandleCache;
         }
 
         /// <summary>
         /// Initialise the dictionary of handles with a list of clusters.
         /// Called at the start to create all our handles in advance.
         /// </summary>
         /// <param name="clusters">List of clusters to get handles for.</param>
-        /// <param name="testing">Testing mode - allows adding non-existent clusters.</param>
-        public static void InitialiseHandles(List<string> clusters, bool testing = false)
+        /// <param name="dideClusters">For testing only - the list of acceptable cluster names.</param>
+        /// <param name="scheduler">For testing only - a scheduler object for mocking.</param>
+        public void InitialiseHandles(List<string> clusters, List<string>? dideClusters = null, IHipercowScheduler? scheduler = null)
         {
             foreach (string cluster in clusters)
             {
-                ClusterHandle.GetClusterHandle(cluster, testing);
+                this.GetClusterHandle(cluster, dideClusters, scheduler);
             }
         }
+
+        /// <summary>
+        /// Return a handle to a named cluster, using the cache where possible,
+        /// creating a new handle if it doesn't exist, or returning null if the
+        /// cluster name was invalid.
+        /// </summary>
+        /// <param name="cluster">The name of the cluster.</param>
+        /// <param name="dideClusters">For testing only - the list of acceptable cluster names.</param>
+        /// <param name="scheduler">For testing only - a scheduler object for mocking.</param>
+        /// <returns>A handle to that cluster, or null if the named cluster does not exist.
+        /// did not exist.</returns>
+        public IHipercowScheduler? GetClusterHandle(string cluster, List<string>? dideClusters = null, IHipercowScheduler? scheduler = null)
+        {
```

Comment:
Could the business with creating the scheduler only if it's null be simplified by making this parameter default to `new HipercowScheduler()`? I *think* that would make you a new scheduler on each call rather than just one that gets shared, but you'd need to check!

Reply:
Yeah, that would be nicer, but it won't let me do it (see below...!)